compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -2027,12 +2027,19 @@ class Typer(@constructorOnly nestingLevel: Int = 0) extends Namer
       var checkedArgs = preCheckKinds(args1, paramBounds)
         // check that arguments conform to bounds is done in phase PostTyper
       val tycon = tpt1.symbol
-      if (tycon == defn.andType)
-        checkedArgs = checkedArgs.mapconserve(arg =>
-          checkSimpleKinded(checkNoWildcard(arg)))
-      else if (tycon == defn.orType)
+      if tycon == defn.andType || tycon == defn.orType then
         checkedArgs = checkedArgs.mapconserve(arg =>
           checkSimpleKinded(checkNoWildcard(arg)))
+      else if tycon.flagsUNSAFE.is(Provisional) then
+        // A type with Provisional flag is either an alias or abstract type.
+        // If it is an alias type, it would mean the type is cyclic
+        // If it is an abstract type, it would mean the type is an irreducible
+        // application of a higher-kinded type to a wildcard argument.
+        // Either way, the wildcard argument is illegal.
+        // The early test here is needed, so that we do not accidentally reduce
```

Comment:
Is the call to `checkNoWildcard` you're adding the early test?  Earlier than what?  I see `preCheckKinds` and the bounds checking in PostTyper, but I can't tell where the accidental reduction can happen.

(Also, there's a typo "roght" if you're pushing another commit.)

Reply:
Yes, `checkNoWildcard`.